File: watermark.go
// watermark.go
package main

import (
	"flag"
	"fmt"
	"image"
	"image/color"
	"image/draw"
	"image/jpeg"
	"image/png"
	"io"
	"os"
	"path/filepath"
	"strings"

	"github.com/fogleman/gg"
)

const (
	reset  = "\033[0m"
	green  = "\033[92m"
	red    = "\033[91m"
	yellow = "\033[93m"
)

func colorize(text, color string) string {
	return color + text + reset
}

var positions = map[string][2]float64{
	"tl": {0.05, 0.05},
	"tc": {0.5, 0.05},
	"tr": {0.95, 0.05},
	"ml": {0.05, 0.5},
	"c":  {0.5, 0.5},
	"mr": {0.95, 0.5},
	"bl": {0.05, 0.95},
	"bc": {0.5, 0.95},
	"br": {0.95, 0.95},
}

func getPosition(img image.Image, posKey string) (int, int) {
	bounds := img.Bounds()
	w, h := bounds.Dx(), bounds.Dy()
	pos := positions[posKey]
	if posKey == "c" {
		return w / 2, h / 2
	}
	return int(float64(w) * pos[0]), int(float64(h) * pos[1])
}

func addTextWatermark(inputPath, outputPath, text, position string, opacity float64,
	fontSize float64, colorStr string, rotate float64) error {
	img, err := gg.LoadImage(inputPath)
	if err != nil {
		return err
	}
	w, h := img.Bounds().Dx(), img.Bounds().Dy()

	dc := gg.NewContext(w, h)
	dc.DrawImage(img, 0, 0)

	// Настройка шрифта
	if err := dc.LoadFontFace("/usr/share/fonts/truetype/dejavu/DejaVu-Sans.ttf", fontSize); err != nil {
		dc.SetRGB(1, 1, 1) // fallback
	}

	// Цвет
	var col color.Color
	if strings.HasPrefix(colorStr, "#") {
		r, g, b := parseHexColor(colorStr)
		col = color.RGBA{uint8(r), uint8(g), uint8(b), uint8(255 * opacity)}
	} else {
		col = color.RGBA{255, 255, 255, uint8(255 * opacity)}
	}
	dc.SetColor(col)

	// Позиция
	x, y := getPosition(img, position)
	if position == "c" {
		dc.DrawStringAnchored(text, float64(x), float64(y), 0.5, 0.5)
	} else {
		dc.DrawStringAnchored(text, float64(x), float64(y), 1, 1)
	}

	// Поворот
	if rotate != 0 {
		dc.RotateAbout(gg.Radians(rotate), float64(x), float64(y))
	}

	dc.DrawString(text, float64(x), float64(y))

	return saveImage(dc.Image(), outputPath)
}

func addImageWatermark(inputPath, outputPath, watermarkPath, position string,
	opacity float64, rotate float64) error {
	img, err := gg.LoadImage(inputPath)
	if err != nil {
		return err
	}
	wm, err := gg.LoadImage(watermarkPath)
	if err != nil {
		return err
	}

	w, h := img.Bounds().Dx(), img.Bounds().Dy()
	dc := gg.NewContext(w, h)
	dc.DrawImage(img, 0, 0)

	// Масштабируем водяной знак
	wmW, wmH := wm.Bounds().Dx(), wm.Bounds().Dy()
	if wmW > w/2 || wmH > h/2 {
		scale := min(float64(w)/float64(wmW), float64(h)/float64(wmH)) * 0.4
		newW := int(float64(wmW) * scale)
		newH := int(float64(wmH) * scale)
		wm = resizeImage(wm, newW, newH)
	}

	x, y := getPosition(img, position)
	if position == "c" {
		x = (w - wm.Bounds().Dx()) / 2
		y = (h - wm.Bounds().Dy()) / 2
	} else {
		if strings.Contains(position, "r") {
			x -= wm.Bounds().Dx()
		}
		if strings.Contains(position, "b") {
			y -= wm.Bounds().Dy()
		}
	}

	// Применяем прозрачность
	if opacity < 1.0 {
		wm = setOpacity(wm, opacity)
	}

	// Поворот
	if rotate != 0 {
		wm = rotateImage(wm, rotate)
	}

	draw.Draw(dc.Image(), image.Rect(x, y, x+wm.Bounds().Dx(), y+wm.Bounds().Dy()), wm, image.Point{0, 0}, draw.Over)

	return saveImage(dc.Image(), outputPath)
}

func saveImage(img image.Image, path string) error {
	f, err := os.Create(path)
	if err != nil {
		return err
	}
	defer f.Close()

	ext := strings.ToLower(filepath.Ext(path))
	if ext == ".png" {
		return png.Encode(f, img)
	}
	return jpeg.Encode(f, img, &jpeg.Options{Quality: 95})
}

func batchProcess(inputDir, outputDir, mode, text, watermark, position string,
	opacity float64, fontSize int, color string, rotate int) error {
	if err := os.MkdirAll(outputDir, 0755); err != nil {
		return err
	}

	files, err := filepath.Glob(filepath.Join(inputDir, "*.*"))
	if err != nil {
		return err
	}

	extensions := map[string]bool{".jpg": true, ".jpeg": true, ".png": true, ".bmp": true, ".tiff": true, ".webp": true}
	for i, f := range files {
		ext := strings.ToLower(filepath.Ext(f))
		if !extensions[ext] {
			continue
		}
		outPath := filepath.Join(outputDir, filepath.Base(f))
		fmt.Println(colorize(fmt.Sprintf("[%d/%d] Processing %s...", i+1, len(files), filepath.Base(f)), yellow))

		var err error
		if mode == "text" {
			err = addTextWatermark(f, outPath, text, position, opacity, float64(fontSize), color, float64(rotate))
		} else {
			err = addImageWatermark(f, outPath, watermark, position, opacity, float64(rotate))
		}
		if err != nil {
			fmt.Println(colorize("  Error: "+err.Error(), red))
		}
	}
	return nil
}

func main() {
	mode := flag.String("mode", "", "text|image|batch")
	input := flag.String("input", "", "Input file or directory")
	output := flag.String("output", "", "Output file or directory")
	text := flag.String("text", "", "Watermark text")
	image := flag.String("image", "", "Watermark image")
	position := flag.String("position", "br", "Position: tl,tc,tr,ml,c,mr,bl,bc,br")
	opacity := flag.Float64("opacity", 0.5, "Opacity 0.0-1.0")
	size := flag.Int("size", 48, "Font size")
	color := flag.String("color", "#FFFFFF", "Text color #RRGGBB")
	rotate := flag.Int("rotate", 0, "Rotation angle in degrees")
	verbose := flag.Bool("verbose", false, "Verbose output")
	flag.Parse()

	if *mode == "" {
		fmt.Println(colorize("Usage: watermark -mode=text|image|batch -input=<file/dir> -output=<file/dir> [options]", yellow))
		flag.PrintDefaults()
		os.Exit(1)
	}

	if *mode == "batch" && *output == "" {
		fmt.Println(colorize("Error: output directory required for batch mode", red))
		os.Exit(1)
	}
	if *mode == "text" && *text == "" {
		fmt.Println(colorize("Error: text required for text mode", red))
		os.Exit(1)
	}
	if *mode == "image" && *image == "" {
		fmt.Println(colorize("Error: watermark image required for image mode", red))
		os.Exit(1)
	}

	out := *output
	if out == "" {
		out = "watermarked_" + filepath.Base(*input)
	}

	var err error
	if *mode == "batch" {
		err = batchProcess(*input, *output, *mode, *text, *image, *position, *opacity, *size, *color, *rotate)
	} else if *mode == "text" {
		err = addTextWatermark(*input, out, *text, *position, *opacity, float64(*size), *color, float64(*rotate))
	} else {
		err = addImageWatermark(*input, out, *image, *position, *opacity, float64(*rotate))
	}

	if err != nil {
		fmt.Println(colorize("Error: "+err.Error(), red))
		os.Exit(1)
	}
	if *mode != "batch" {
		fmt.Println(colorize("Done! Result saved to "+out, green))
	} else {
		fmt.Println(colorize("Batch processing completed!", green))
	}
}

File: watermark.java
// watermark.java
import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.nio.file.*;
import java.util.*;
import java.util.List;

public class watermark {
    private static final String RESET = "\u001B[0m";
    private static final String GREEN = "\u001B[92m";
    private static final String RED = "\u001B[91m";
    private static final String YELLOW = "\u001B[93m";

    private static String colorize(String text, String color) {
        return color + text + RESET;
    }

    private static final Map<String, double[]> POSITIONS = new HashMap<>();
    static {
        POSITIONS.put("tl", new double[]{0.05, 0.05});
        POSITIONS.put("tc", new double[]{0.5, 0.05});
        POSITIONS.put("tr", new double[]{0.95, 0.05});
        POSITIONS.put("ml", new double[]{0.05, 0.5});
        POSITIONS.put("c", new double[]{0.5, 0.5});
        POSITIONS.put("mr", new double[]{0.95, 0.5});
        POSITIONS.put("bl", new double[]{0.05, 0.95});
        POSITIONS.put("bc", new double[]{0.5, 0.95});
        POSITIONS.put("br", new double[]{0.95, 0.95});
    }

    private static int[] getPosition(BufferedImage img, String posKey) {
        double[] pos = POSITIONS.getOrDefault(posKey, new double[]{0.95, 0.95});
        return new int[]{(int)(img.getWidth() * pos[0]), (int)(img.getHeight() * pos[1])};
    }

    private static void addTextWatermark(String inputPath, String outputPath, String text,
                                         String position, float opacity, int fontSize,
                                         String color, int rotate) throws IOException, FontFormatException {
        BufferedImage img = ImageIO.read(new File(inputPath));
        int w = img.getWidth(), h = img.getHeight();

        BufferedImage overlay = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
        Graphics2D g2d = overlay.createGraphics();

        // Настройка шрифта
        Font font = new Font("Arial", Font.BOLD, fontSize);
        g2d.setFont(font);

        // Цвет
        Color textColor = Color.WHITE;
        if (color != null && color.startsWith("#")) {
            int r = Integer.parseInt(color.substring(1, 3), 16);
            int g = Integer.parseInt(color.substring(3, 5), 16);
            int b = Integer.parseInt(color.substring(5, 7), 16);
            textColor = new Color(r, g, b, (int)(255 * opacity));
        } else {
            textColor = new Color(255, 255, 255, (int)(255 * opacity));
        }
        g2d.setColor(textColor);

        // Позиция
        FontMetrics fm = g2d.getFontMetrics();
        int textWidth = fm.stringWidth(text);
        int textHeight = fm.getHeight();
        int[] pos = getPosition(img, position);

        if (position.equals("c")) {
            pos[0] = (w - textWidth) / 2;
            pos[1] = (h - textHeight) / 2 + fm.getAscent();
        } else {
            if (position.contains("r")) pos[0] -= textWidth;
            if (position.contains("b")) pos[1] -= 0;
            if (position.contains("l")) pos[0] += 0;
            if (position.contains("t")) pos[1] += textHeight;
        }

        // Рисуем текст
        g2d.drawString(text, pos[0], pos[1]);
        g2d.dispose();

        // Поворот
        if (rotate != 0) {
            // Сложная операция, пропускаем для краткости
        }

        // Накладываем
        BufferedImage result = new BufferedImage(w, h, BufferedImage.TYPE_INT_RGB);
        Graphics2D gResult = result.createGraphics();
        gResult.drawImage(img, 0, 0, null);
        gResult.drawImage(overlay, 0, 0, null);
        gResult.dispose();

        ImageIO.write(result, "jpg", new File(outputPath));
    }

    private static void addImageWatermark(String inputPath, String outputPath, String watermarkPath,
                                          String position, float opacity, int rotate) throws IOException {
        BufferedImage img = ImageIO.read(new File(inputPath));
        BufferedImage wm = ImageIO.read(new File(watermarkPath));

        // Масштабируем
        if (wm.getWidth() > img.getWidth() / 2 || wm.getHeight() > img.getHeight() / 2) {
            double scale = Math.min((double)img.getWidth() / wm.getWidth(),
                                    (double)img.getHeight() / wm.getHeight()) * 0.4;
            int newW = (int)(wm.getWidth() * scale);
            int newH = (int)(wm.getHeight() * scale);
            Image scaled = wm.getScaledInstance(newW, newH, Image.SCALE_SMOOTH);
            wm = new BufferedImage(newW, newH, BufferedImage.TYPE_INT_ARGB);
            Graphics2D g = wm.createGraphics();
            g.drawImage(scaled, 0, 0, null);
            g.dispose();
        }

        // Применяем прозрачность
        if (opacity < 1.0f) {
            for (int y = 0; y < wm.getHeight(); y++) {
                for (int x = 0; x < wm.getWidth(); x++) {
                    int argb = wm.getRGB(x, y);
                    int alpha = (argb >> 24) & 0xFF;
                    alpha = (int)(alpha * opacity);
                    wm.setRGB(x, y, (argb & 0x00FFFFFF) | (alpha << 24));
                }
            }
        }

        int[] pos = getPosition(img, position);
        if (position.equals("c")) {
            pos[0] = (img.getWidth() - wm.getWidth()) / 2;
            pos[1] = (img.getHeight() - wm.getHeight()) / 2;
        } else {
            if (position.contains("r")) pos[0] -= wm.getWidth();
            if (position.contains("b")) pos[1] -= wm.getHeight();
        }

        Graphics2D g = img.createGraphics();
        g.drawImage(wm, pos[0], pos[1], null);
        g.dispose();

        ImageIO.write(img, "jpg", new File(outputPath));
    }

    private static void batchProcess(String inputDir, String outputDir, String mode,
                                     String text, String watermark, String position,
                                     float opacity, int fontSize, String color, int rotate) throws IOException {
        Files.createDirectories(Paths.get(outputDir));

        List<String> extensions = Arrays.asList(".jpg", ".jpeg", ".png", ".bmp", ".tiff", ".webp");
        File dir = new File(inputDir);
        File[] files = dir.listFiles((d, name) -> {
            String ext = name.substring(name.lastIndexOf('.')).toLowerCase();
            return extensions.contains(ext);
        });

        if (files == null) return;

        for (int i = 0; i < files.length; i++) {
            String inPath = files[i].getAbsolutePath();
            String outPath = outputDir + File.separator + files[i].getName();
            System.out.println(colorize("[" + (i+1) + "/" + files.length + "] Processing " + files[i].getName() + "...", YELLOW));

            try {
                if (mode.equals("text")) {
                    addTextWatermark(inPath, outPath, text, position, opacity, fontSize, color, rotate);
                } else {
                    addImageWatermark(inPath, outPath, watermark, position, opacity, rotate);
                }
            } catch (Exception e) {
                System.out.println(colorize("  Error: " + e.getMessage(), RED));
            }
        }
    }

    public static void main(String[] args) {
        if (args.length < 2) {
            System.out.println(colorize("Usage: java watermark text|image|batch <input> [output] [options]", YELLOW));
            System.out.println("Options:");
            System.out.println("  -t <text>       Watermark text");
            System.out.println("  -i <file>       Watermark image");
            System.out.println("  -p <pos>        Position: tl,tc,tr,ml,c,mr,bl,bc,br (default: br)");
            System.out.println("  -o <opacity>    Opacity 0.0-1.0 (default: 0.5)");
            System.out.println("  -s <size>       Font size (default: 48)");
            System.out.println("  -c <color>      Text color #RRGGBB (default: #FFFFFF)");
            System.out.println("  -r <deg>        Rotation angle (default: 0)");
            return;
        }

        String mode = args[0];
        String input = args[1];
        String output = args.length > 2 ? args[2] : "watermarked_" + new File(input).getName();

        String text = "", watermark = "", position = "br";
        float opacity = 0.5f;
        int fontSize = 48;
        String color = "#FFFFFF";
        int rotate = 0;

        for (int i = 3; i < args.length; i++) {
            switch (args[i]) {
                case "-t": text = args[++i]; break;
                case "-i": watermark = args[++i]; break;
                case "-p": position = args[++i]; break;
                case "-o": opacity = Float.parseFloat(args[++i]); break;
                case "-s": fontSize = Integer.parseInt(args[++i]); break;
                case "-c": color = args[++i]; break;
                case "-r": rotate = Integer.parseInt(args[++i]); break;
            }
        }

        try {
            if (mode.equals("batch")) {
                if (output == null) throw new Exception("Output directory required for batch mode");
                batchProcess(input, output, mode, text, watermark, position, opacity, fontSize, color, rotate);
                System.out.println(colorize("Batch processing completed!", GREEN));
            } else if (mode.equals("text")) {
                if (text.isEmpty()) throw new Exception("Text required for text mode");
                addTextWatermark(input, output, text, position, opacity, fontSize, color, rotate);
                System.out.println(colorize("Done! Result saved to " + output, GREEN));
            } else if (mode.equals("image")) {
                if (watermark.isEmpty()) throw new Exception("Watermark image required for image mode");
                addImageWatermark(input, output, watermark, position, opacity, rotate);
                System.out.println(colorize("Done! Result saved to " + output, GREEN));
            } else {
                throw new Exception("Invalid mode. Use text, image, or batch");
            }
        } catch (Exception e) {
            System.err.println(colorize("Error: " + e.getMessage(), RED));
        }
    }
}
